crates/sui-core/src/execution_scheduler/balance_withdraw_scheduler/scheduler.rs

```diff
@@ -0,0 +1,156 @@
+    balance_read::AccountBalanceRead, naive_scheduler::NaiveBalanceWithdrawScheduler,
+    BalanceSettlement, ScheduleResult, TxBalanceWithdraw,
+};
+use mysten_metrics::monitored_mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use sui_types::{base_types::SequenceNumber, digests::TransactionDigest};
+use tokio::sync::oneshot;
+
+#[allow(dead_code)]
+#[async_trait::async_trait]
+pub(crate) trait BalanceWithdrawSchedulerTrait: Send + Sync {
+    async fn schedule_withdraws(&self, withdraws: WithdrawReservations);
+    async fn settle_balances(&self, settlement: BalanceSettlement);
+}
+
+#[allow(dead_code)]
+pub(crate) struct WithdrawReservations {
+    pub accumulator_version: SequenceNumber,
+    pub withdraws: Vec<TxBalanceWithdraw>,
+    pub senders: Vec<oneshot::Sender<ScheduleResult>>,
+}
+
+#[allow(dead_code)]
+#[derive(Clone)]
+pub(crate) struct BalanceWithdrawScheduler {
+    inner: Arc<dyn BalanceWithdrawSchedulerTrait>,
+    /// Use channels to process withdraws and settlements asynchronously without blocking the caller.
+    withdraw_sender: UnboundedSender<WithdrawReservations>,
+    settlement_sender: UnboundedSender<BalanceSettlement>,
+}
+
+impl WithdrawReservations {
+    #[allow(dead_code)]
+    pub fn new(
+        accumulator_version: SequenceNumber,
+        withdraws: Vec<TxBalanceWithdraw>,
+    ) -> (
+        Self,
+        BTreeMap<TransactionDigest, oneshot::Receiver<ScheduleResult>>,
+    ) {
+        let (senders, receivers) = withdraws
+            .iter()
+            .map(|withdraw| {
+                let (sender, receiver) = oneshot::channel();
+                (sender, (withdraw.tx_digest, receiver))
+            })
+            .unzip();
+        (
+            Self {
+                accumulator_version,
+                withdraws,
+                senders,
+            },
+            receivers,
+        )
+    }
+}
+
+impl BalanceWithdrawScheduler {
+    #[allow(dead_code)]
+    pub fn new(
+        balance_read: Arc<dyn AccountBalanceRead>,
+        init_accumulator_version: SequenceNumber,
+    ) -> Arc<Self> {
+        let inner = NaiveBalanceWithdrawScheduler::new(balance_read, init_accumulator_version);
+        let (withdraw_sender, withdraw_receiver) =
+            unbounded_channel("withdraw_scheduler_withdraws");
+        let (settlement_sender, settlement_receiver) =
+            unbounded_channel("withdraw_scheduler_settlements");
+        let scheduler = Arc::new(Self {
+            inner,
+            withdraw_sender,
+            settlement_sender,
+        });
+        tokio::spawn(scheduler.clone().process_withdraw_task(withdraw_receiver));
+        tokio::spawn(
+            scheduler
+                .clone()
+                .process_settlement_task(settlement_receiver, init_accumulator_version),
+        );
+        scheduler
+    }
+
+    /// This function will be called either by ConsensusHandler or the CheckpointExecutor.
```

Comment:
i don't think it should be called by CheckpointExecutor, right? Validators have to do their own scheduling/settlement decisions.

Reply:
It would have to be.
On a fullnode, you also want optimistic withdraw scheduling, otherwise execution becomes very sequential.
The implementation explicitly tolerates this to be called from both places.